### PR TITLE
rune pal - Add a base support for aws sample nitro enclave

### PR DESCRIPTION
**Pull request by Simon-aikuier**
Implement a pal interface to support was nitro enclave.
For now just to run the sample enclave image file in EC2 instance.

### CODE REVIEW

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c [hunk 1]
```diff
@@ -0,0 +1,829 @@
...
+ * @slot_uid:		The generated slot uid for the enclave.
+ * @enclave_fd :	The generated file descriptor for the enclave.
+ *
+ * Context: Process context.
+ * Return:
+ * * 0 on success.
+ * * Negative return value on failure.
+ */
+static int ne_create_vm(int ne_dev_fd, unsigned long *slot_uid, int *enclave_fd)
+{
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c:102]: Change to:
```
{
  *enclave_fd = ioctl(ne_dev_fd, NE_CREATE_VM, slot_uid);
  if (*enclave_fd >= 0)
    return 0;

  // stuffs about *enclave_fd < 0
}
```

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c [hunk 2]
```diff
@@ -0,0 +1,829 @@
...
+			       i * NE_POLL_WAIT_TIME);
+
+			continue;
+		}
+
+		printf("Poll received value 0x%x\n", fds[0].revents);
+
+		if (fds[0].revents & POLLHUP) {
+			printf("Received POLLHUP\n");
+
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c:169]: remote this newline.

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+			return NULL;
+		}
+
+		i++;
+
+		if (!rc) {
+			printf("Poll: %d seconds elapsed\n",
+			       i * NE_POLL_WAIT_TIME);
+
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c:161]: remote this newline.

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+		printf("Error in the enclave start / image loading heartbeat logic [rc=%d]\n", rc);
+
+		goto release_enclave_fd;
+	}
+
+	return 0;
+
+release_enclave_fd:
+	close(enclave_fd);
+	ne_free_mem_regions(ne_user_mem_regions);
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c:868]: enclave_fd should be closed after ne_free_mem_regions(ne_user_mem_regions)

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c [hunk 5]
```diff
@@ -0,0 +1,829 @@
...
+
+int pal_kill(int pid, int sig)
+{
+	return 0;
+}
+
+int pal_destroy(void)
+{
+	close(enclave_fd);
+	ne_free_mem_regions(ne_user_mem_regions);
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/pal_ne.c:892]: ditto

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/README.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
+The files in this directory are used to implement a nitro enclave runtime in order to help to write your own enclave runtime. For now just implement the pal interface for nitro enclave to run the sample enclave image file.
+
+---
+
+# Run nitro enclave with OCI bundle
+## Create EC2 instance with Enclave enable
+refer to [Amazon EC2 instances](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Instances.html) to create a `EC2` instance which need to enable Enclave. Note that the limitation:
```
Comment (haosanzi) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/README.md:7]: s/refer/Refer/.

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/README.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+- virtualized Nitro-based instances with at least 4 vCPUs, except t3, t3a, t4g, a1, c6g, c6gd, m6g, m6gd, r6g, and r6gd.
+- Parent instance and Enclave both are linux operating system.
+
+## Build nitro enclave environment
+refer to [Getting started: Hello enclave](https://docs.aws.amazon.com/enclaves/latest/user/getting-started.html) to build a `nitro enclave` environment in parent instance.
```
Comment (haosanzi) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/README.md:12]: s/refer/Refer/.

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/ne.c [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+ * @enclave_fd :	The generated file descriptor for the enclave.
+ *
+ * Context: Process context.
+ * Return:
+ * * 0 on success.
+ * * Negative return value on failure.
+ */
+int ne_create_vm(int ne_dev_fd, unsigned long *slot_uid, int *enclave_fd)
+{
+	int rc = -EINVAL;
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/ne.c:36]: Why still leaving this line?

File: rune/libenclave/internal/runtime/pal/nitro_enclaves/ne.c [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+	*enclave_fd = ioctl(ne_dev_fd, NE_CREATE_VM, slot_uid);
+	if (*enclave_fd >= 0)
+		return 0;
+
+	rc = *enclave_fd;
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/nitro_enclaves/ne.c:41]: Why returning enclave fd instead of -EINVAL?

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Comment by jiazhang0** [comment]
Closed. Review new pr https://github.com/alibaba/inclavare-containers/pull/354

**Pull request closed without merging**